File: week15/보연/PG_도넛과막대그래프.cpp
#include <cstring>
#include <string>
#include <vector>
#include <iostream>
using namespace std;

const int SIZE = 1000001;
int N;
vector<int> nodeList[SIZE];
bool visited[SIZE];
int donut, line, eight;
int in_degree[SIZE];
int out_degree[SIZE];

void dfs(int start, int curr) {
    if(out_degree[curr] == 0) {
        line++;
        return;
    }
    if(out_degree[curr] >= 2) {
        eight++;
        return;
    }
    if(visited[curr]) {
        donut++;
        return;
    }
    visited[curr] = true;
    
    for(int i=0; i<nodeList[curr].size(); i++) {
        int next = nodeList[curr][i];
        dfs(start, next);
    }
}

vector<int> solution(vector<vector<int>> edges) {
    vector<int> answer;
    N = edges.size();
    for(int i=0; i<edges.size(); i++) {
        int u = edges[i][0];
        int v = edges[i][1];
        
        nodeList[u].push_back(v);
        in_degree[v] ++;
        out_degree[u] ++;
    }
    int root = 0;
    for(int i=0; i<N; i++) {
        if(in_degree[i] == 0 && out_degree[i] >= 2) {
            root = i;
            break;
        }
    }
    
    memset(visited, 0, sizeof(visited));
    donut=0, line=0, eight=0;
    for(int i=0; i<nodeList[root].size(); i++) {
        dfs(nodeList[root][i], nodeList[root][i]);
    }
    answer.push_back(root);
    answer.push_back(donut);
    answer.push_back(line);
    answer.push_back(eight);

    return answer;
}

File: week15/지환/PG_도넛과막대그래프.java
import java.util.*;

class PG_도넛과막대그래프 {
    static int[] in = new int[1000001];
    static int[] out = new int[1000001];
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];
        for(int[] edge: edges) {
            in[edge[1]]++;
            out[edge[0]]++;
        }
        
         for(int i=1; i<=1000000; i++) {
            if(in[i] == 0 && out[i] >= 2) {
                answer[0] = i;
            } else if (in[i] > 0 && out[i] == 0) {
                answer[2]++;
            } else if (in[i] >= 2 && out[i] == 2) {
                answer[3]++;
            }
        }
        int cnt = out[answer[0]];
        answer[1] = cnt - answer[2] - answer[3];
        return answer;
    }
    
}
